diff -Naurb SOAPBuilder.py SOAPBuilder.py
--- SOAPBuilder.py	2008-12-22 14:42:14.000000000 -0500
+++ SOAPBuilder.py	2011-10-05 11:15:33.454752000 -0400
@@ -134,8 +134,10 @@
             else:
                 methodns, n = '', ''

-            self.out.append('<%s%s%s%s%s>\n' % (
+            self.out.append('<%s%s%s%s%s>' % (
                 methodns, self.method, n, a, self.genroot(ns_map)))
+            if self.args or self.kw:
+              self.out.append('\n')

         try:
             if type(self.args) != TupleType:
@@ -579,13 +581,13 @@
             # first write out items with order information
             if hasattr(obj, '_keyord'):
                 for i in range(len(obj._keyord)):
-                    self.dump(obj._aslist(i), obj._keyord[i], 1, ns_map)
+                    self.dump(obj._aslist(i), obj._keyord[i], typed, ns_map)
                     keylist.remove(obj._keyord[i])

             # now write out the rest
             for k in keylist:
                 if (k[0] != "_"):
-                    self.dump(getattr(obj,k), k, 1, ns_map)
+                    self.dump(getattr(obj,k), k, typed, ns_map)

             if isinstance(obj, bodyType):
                 self.multis = 1
diff -Naurb wstools/Namespaces.py wstools/Namespaces.py
--- wstools/Namespaces.py	2008-12-22 14:42:13.000000000 -0500
+++ wstools/Namespaces.py	2011-10-05 09:09:17.811673000 -0400
@@ -122,4 +122,7 @@
     CORE    = "http://www.globus.org/namespaces/2004/06/core"
     SIG     = "http://www.globus.org/2002/04/xmlenc#gssapi-sign"

+class JAXB:
+    NS = "http://java.sun.com/xml/ns/jaxb"
+
 ZSI_SCHEMA_URI = 'http://www.zolera.com/schemas/ZSI/'
diff -Naurb wstools/XMLSchema.py wstools/XMLSchema.py
--- wstools/XMLSchema.py	2005-02-18 08:50:14.000000000 -0500
+++ wstools/XMLSchema.py	2011-10-10 10:42:57.382400000 -0400
@@ -16,6 +16,7 @@

 import types, weakref, urllib, sys
 from threading import RLock
+from Namespaces import JAXB
 from Namespaces import XMLNS
 from Utility import DOM, DOMException, Collection, SplitQName
 from StringIO import StringIO
@@ -674,7 +675,7 @@
                     'class instance %s, missing required attribute %s'\
                     %(self.__class__, a)
         for a in self.attributes.keys():
-            if (a not in (XMLSchemaComponent.xmlns, XMLNS.XML)) and\
+            if (a not in (XMLSchemaComponent.xmlns, XMLNS.XML, JAXB.NS)) and\
                 (a not in self.__class__.attributes.keys()) and not\
                 (self.isAttribute() and self.isReference()):
                 raise SchemaError, '%s, unknown attribute(%s,%s)' \
@@ -2860,7 +2861,7 @@
     """Tuple of length 2, consisting of
        a namespace and unprefixed name.
     """
-    def __init__(self, args):
+    def __new__(self, args):
         """args -- (namespace, name)
            Remove the name's prefix, irrelevant.
         """
@@ -2868,8 +2869,7 @@
             raise TypeError, 'expecting tuple (namespace, name), got %s' %args
         elif args[1].find(':') >= 0:
             args = (args[0], SplitQName(args[1])[1])
-        tuple.__init__(self, args)
-        return
+        return tuple.__new__(TypeDescriptionComponent, args)

     def getTargetNamespace(self):
         return self[0]
diff -Naurb Client.py Client.py
--- Client.py	2008-12-22 14:42:14.000000000 -0500
+++ Client.py	2011-10-24 13:55:43.705887000 -0400
@@ -40,11 +40,11 @@
 ################################################################################
 """
 
+from __future__ import nested_scopes
+
 ident = '$Id: Client.py,v 1.27 2005/02/21 20:27:09 warnes Exp $'
 from version import __version__
 
-from __future__ import nested_scopes
-
 #import xml.sax
 import urllib
 from types import *
@@ -112,6 +112,9 @@
 
 
 class HTTPTransport:
+    def __init__(self, additional_headers = None):
+        self.additional_headers = additional_headers or {}
+
     def getNS(self, original_namespace, data):
         """Extract the (possibly extended) namespace from the returned
         SOAP message."""
@@ -153,37 +156,40 @@
 
         r.putrequest("POST", real_path)
 
-        r.putheader("Host", addr.host)
-        r.putheader("User-agent", SOAPUserAgent())
+        headers = []
+
+        headers.append(("Host", addr.host))
+        headers.append(("User-agent", SOAPUserAgent()))
         t = 'text/xml';
         if encoding != None:
             t += '; charset="%s"' % encoding
-        r.putheader("Content-type", t)
-        r.putheader("Content-length", str(len(data)))
+        headers.append(("Content-type", t))
+        headers.append(("Content-length", str(len(data))))
 
         # if user is not a user:passwd format
         #    we'll receive a failure from the server. . .I guess (??)
         if addr.user != None:
             val = base64.encodestring(addr.user) 
-            r.putheader('Authorization','Basic ' + val.replace('\012',''))
+            headers.append(('Authorization','Basic ' + val.replace('\012','')))
+
+        headers.extend(self.additional_headers.items())
 
         # This fixes sending either "" or "None"
         if soapaction == None or len(soapaction) == 0:
-            r.putheader("SOAPAction", "")
+            headers.append(("SOAPAction", ""))
         else:
-            r.putheader("SOAPAction", '"%s"' % soapaction)
+            headers.append(("SOAPAction", '"%s"' % soapaction))
 
         if config.dumpHeadersOut:
             s = 'Outgoing HTTP headers'
             debugHeader(s)
             print "POST %s %s" % (real_path, r._http_vsn_str)
-            print "Host:", addr.host
-            print "User-agent: SOAPpy " + __version__ + " (http://pywebsvcs.sf.net)"
-            print "Content-type:", t
-            print "Content-length:", len(data)
-            print 'SOAPAction: "%s"' % soapaction
+            for header in headers:
+                print '%s:%s' % header
             debugFooter(s)
 
+        for header in headers:
+            r.putheader(header[0], header[1])
         r.endheaders()
 
         if config.dumpSOAPOut:
@@ -493,3 +499,6 @@
 
         def __repr__(self):
             return "<%s at %d>" % (self.__class__, id(self))
+
+        def __str__(self):
+            return repr(self)
diff -Naurb Types.py Types.py
--- Types.py	2009-05-13 18:20:55.000000000 -0400
+++ Types.py	2011-11-29 10:33:00.134756000 -0500
@@ -33,11 +33,11 @@
 ################################################################################
 """
 
+from __future__ import nested_scopes
+
 ident = '$Id: Types.py,v 1.19 2005/02/22 04:29:43 warnes Exp $'
 from version import __version__
 
-from __future__ import nested_scopes
-
 import UserList
 import base64
 import cgi
diff -Naurb Server.py Server.py
--- Server.py	2009-05-13 18:20:55.000000000 -0400
+++ Server.py	2011-11-29 10:33:00.134756000 -0500
@@ -40,11 +40,11 @@
 ################################################################################
 """
 
+from __future__ import nested_scopes
+
 ident = '$Id: Server.py,v 1.21 2005/02/15 16:32:22 warnes Exp $'
 from version import __version__
 
-from __future__ import nested_scopes
-
 #import xml.sax
 import re
 import socket
